File: Assignment1/src/main/kotlin/PostfixCalculator.kt
import java.lang.IllegalArgumentException
import kotlin.math.pow

/*
 * UCSC CSE118 Mobile Applications - Kevin Crawford
 * Made with assistance from: https://rosettacode.org/wiki/Parsing/RPN_calculator_algorithm#Kotlin
 *
 * Assignment 1 : Multi operator, multi operand postfix (Reverse Polish
 * Notation) calculator supporting addition, subtraction, multiplication,
 * division, and power.
 *
 * See assignment specification for details.
 *
 * Copyright (C) 2022 David C. Harrison
 */
class PostfixCalculator {
  /**
   * See assignment specification.
   */
  fun parse(expression: String): Double {
    val operands: List<String> = expression.split(" ")
    val stack: MutableList<Double> = mutableListOf()
    for (i in operands.indices) {
      val operand = operands[i].toDoubleOrNull()
      if (operand == null) {
        if (operands[i] == "+" || operands[i] == "-" ||
          operands[i] == "*" || operands[i] == "/" ||
          operands[i] == "^"
        ) {
          if (stack.size < 2) {
            throw IllegalArgumentException("Stack too small")
          } else {
            val operand2: Double = stack.removeAt(stack.lastIndex)
            val operand1: Double = stack.removeAt(stack.lastIndex)
            if (operands[i] == "+") {
              stack.add(operand1 + operand2)
            } else if (operands[i] == "-") {
              stack.add(operand1 - operand2)
            } else if (operands[i] == "*") {
              stack.add(operand1 * operand2)
            } else if (operands[i] == "/") {
              stack.add(operand1 / operand2)
            } else {
              stack.add(operand1.pow(operand2))
            }
          }
        } else {
          throw IllegalArgumentException("Unknown Operator")
        }
      } else {
        stack.add(operand)
      }
    }
    return stack.elementAt(0)
  }
}

File: Assignment4/Assignment4/Logic/PostfixCalculator.swift
import Foundation

class PostfixCalculator {
  func parse(expression: String) throws -> Double {
    let str = expression.trimmingCharacters(in: .whitespacesAndNewlines)
    let input = str.components(separatedBy: " ")
    var stack = Array<Double>()
    if (input.count < 3) {
      throw PostfixCalculatorError.invalidExpression("Insuffient Operands")
    }
      for i in input.indices {
          let operand = Double(input[i])
          if (operand == nil) {
              if (input[i] == "+" || input[i] == "-" ||
              input[i] == "*" || input[i] == "/" ||
                  input[i] == "^") {
                  if (stack.count < 2) {
                      throw PostfixCalculatorError.invalidExpression("Insufficient Operands")
                  } else {
                      let operand2 = stack.removeLast()
                      let operand1 = stack.removeLast()
                      if (input[i] == "+") {
                        stack.append(operand1 + operand2)
                      } else if (input[i] == "-") {
                        stack.append(operand1 - operand2)
                      } else if (input[i] == "*") {
                        stack.append(operand1 * operand2)
                      } else if (input[i] == "/") {
                        stack.append(operand1 / operand2)
                      } else {
                        stack.append(pow(operand1,operand2))
                      }
                  }
              } else {
                  throw PostfixCalculatorError.invalidOperator("Unknown Operator")
              }
          } else {
              stack.append(operand!)
          }
      }
    return stack[0]
  }
}
